File: Image.java
import java.io.FileWriter;
import java.io.IOException;
import java.util.Scanner;

public class Image {
	
	private int numRows;
	private int numCols;
	private int minValue;
	private int maxValue;
	public int[] histogramArray;
	public int thresholdValue;
	
	Image(Scanner input, int thresholdValue){
		if(input.hasNext()) {
			this.numRows = input.nextInt();
		}
		if(input.hasNext()) {
			this.numCols = input.nextInt();
		}
		if(input.hasNext()) {
			this.minValue = input.nextInt();
		}
		if(input.hasNext()) {
			this.maxValue = input.nextInt();
		}
		this.thresholdValue = thresholdValue;
		
		this.histogramArray = new int[maxValue + 1];
		
		for(int i = 0; i <= maxValue; i++) {
			histogramArray[i] = 0;
		}
	}
	
	public void computeHistogram(Scanner input, int[] array) {
		int value = -1;
		while(input.hasNext()) {
			value = input.nextInt();
			array[value]++;
		}		
	}
	
	public void printHistogram(FileWriter output) {
		String str = "";
		try {
			output.write(numRows + " " + numCols + " " + minValue + " " + maxValue + "\n");
			for(int i = 0; i <= maxValue; i++) {
				
				str = "";
				if(i <= 9) {
					str += "0";
				}
				
				str += i + "  (";
				if(histogramArray[i] < 10) {
					str += "00";
				}
				else if(histogramArray[i] < 100 && histogramArray[i] > 10) {
					str += "0";
				}
				str += histogramArray[i] + "):";
				
				if(histogramArray[i] != 0) {
					int max = histogramArray[i];
					if(max > 60) {
						max = 60;
					}
					for(int j = 0; j < max; j++) {
						str += "+";
					}
				}
				output.write(str + "\n");
			}			
		}
		catch(IOException e) {
			e.printStackTrace();
		}
	}
	
	public void thresholdOperation(Scanner input, int threshold, FileWriter output) {
		int row = 0, col = 0, pixelValue = -1;
		try {
			output.write(numRows + " " + numCols + " " + minValue + " " + maxValue + "\n");
			
			if(input.hasNext()) {
				this.numRows = input.nextInt();
			}
			if(input.hasNext()) {
				this.numCols = input.nextInt();
			}
			if(input.hasNext()) {
				this.minValue = input.nextInt();
			}
			if(input.hasNext()) {
				this.maxValue = input.nextInt();
			}
			
			while(row < numRows) {
				col = 0;
				while(col < numCols) {
					if(input.hasNext()) {
						pixelValue = input.nextInt();
					}
					else {
						break;
					}
					if(pixelValue >= threshold) {
						output.write("1 ");
					}
					else {
						output.write("0 ");
					}
					col++;
				}
				output.write("\n");
				row++;
			}
		}
		catch(IOException e) {
			e.printStackTrace();
		}
	}
	
	public void prettyPrint(Scanner input, FileWriter output) {
		int row = 0, col = 0, pixelValue = -1;
		try {
			
			if(input.hasNext()) {
				this.numRows = input.nextInt();
			}
			if(input.hasNext()) {
				this.numCols = input.nextInt();
			}
			if(input.hasNext()) {
				this.minValue = input.nextInt();
			}
			if(input.hasNext()) {
				this.maxValue = input.nextInt();
			}
			
			while(row < numRows) {
				col = 0;
				while(col < numCols) {
					if(input.hasNext()) {
						pixelValue = input.nextInt();
					}
					else {
						break;
					}
					if(pixelValue > 0) {
						output.write(pixelValue  + " ");
					}
					else {
						output.write("  ");
					}
					col++;
				}
				output.write("\n");
				row++;
			}
		}
		catch(IOException e) {
			e.printStackTrace();
		}
	}

}

File: C++/Image.cpp
#include "Image.h"

Image::Image(ifstream& input, int thresholdValue){
	if (input.good()) {
		input >> this->numRows;
		input >> this->numCols;
		input >> this->minValue;
		input >> this->maxValue;
	}
	this->thresholdValue = thresholdValue;
	histogramArray = new int[this->maxValue + 1];

	for (int i = 0; i <= this->maxValue; i++) {
		histogramArray[i] = 0;
	}
}

void Image::computeHistogram(ifstream& input, int* array){
	int value = -1;
	while (!input.eof()) {
		input >> value;
		histogramArray[value]++;
	}
}

void Image::printHistogram(ofstream& output){
	try {
		for (int i = 0; i <= this->maxValue; i++) {
			if (i <= 9) {
				output << "0";
			}
			
			output << i << "  (";
			if (histogramArray[i] < 10) {
				output << "00";
			}
			else if (histogramArray[i] < 100 && histogramArray[i] > 9) {
				output << "0";
			}
			output << histogramArray[i] << "):";
			if (histogramArray[i] != 0) {
				int max = histogramArray[i];
				if (max > 60) {
					max = 60;
				}
				for (int j = 0; j < max; j++) {
					output << "+";
				}
			}
			output << endl;

		}

	}
	catch (exception e) {
		cout << "There is an error: " << e.what() << endl;
	}
}

void Image::thresholdOperation(ifstream& input, int threshold, ofstream& output) {
	int row = 0, col = 0, pixelValue = -1;
	try {
		output << numRows << " " << numCols << " " << minValue << " " << maxValue << endl;
		if(input.good()){
			input >> numRows;
			input >> numCols;
			input >> minValue;
			input >> maxValue;
		}

		while (row < numRows) {
			col = 0;
			while (col < numCols) {
				if (input.good()) {
					input >> pixelValue;
				}
				else {
					break;
				}
				if (pixelValue >= threshold) {
					output << "1 ";
				}
				else {
					output << "0 ";
				}
				col++;
			}
			output << endl;
			row++;
		}
	}
	catch (exception e) {
		cout << "There is an error: " << e.what() << endl;
	}

}

void Image::prettyPrint(ifstream& input, ofstream& output){
	int row = 0, col = 0, pixelValue = -1;
	try {
		output << numRows << " " << numCols << " " << minValue << " " << maxValue << endl;
		if (input.good()) {
			input >> numRows;
			input >> numCols;
			input >> minValue;
			input >> maxValue;
		}

		while (row < numRows) {
			col = 0;
			while (col < numCols) {
				if (input.good()) {
					input >> pixelValue;
				}
				else {
					break;
				}
				if (pixelValue > 0) {
					output << pixelValue << " ";
				}
				else {
					output << "  ";
				}
				col++;
			}
			output << endl;
			row++;
		}
	}
	catch (exception e) {
		cout << "There is an error: " << e.what() << endl;
	}
}
